tests/test_google_analytics_bookmarks.py

```diff
@@ -195,22 +190,19 @@ def test_run(self):
                     second_replication_values_sorted = sorted(second_replication_values)
                     self.assertEqual(second_replication_values_sorted, second_replication_values)
 
-                    # Verify the second sync records respect the previous (simulated) bookmark value
+                   
                     for record in second_sync_messages:
                         with self.subTest(record=record):
+                            # Verify the second sync records respect the previous (simulated) bookmark value
                             self.assertGreaterEqual(record[replication_key], simulated_bookmark_value)
 
-                    # UNCOMMENT START BUG_TDL-16943
```

Comment:
@prijendev Can you please reply to this thread what change you are commenting out?

Reply:
Here, the assertion was verifying that a sync's bookmarked value was equal to today's date for a report using incremental replication. But, this is invalid because google-analytics bookmarks based on the last record which will not always be for today. So, removed these assertions.

Reply:
Could you please provide documentation from google or point to the code to prove that this assertion is invalid. At the time of writing this test this was a requirement. I am not convinced that just because the tap does not replicate through to today's date, that means google-analytics is not returning data for the current day.

Reply:
@kspeer825 Actually, this same issue was mentioned in [#58 ](https://github.com/singer-io/tap-google-analytics/pull/58) But it seems like, the expected assertion was removed instead of an unexpected assertion. 
Found one online link, which says it can take between 24 and 48 hours for Google Analytics to finish processing data.  [Link](https://stackoverflow.com/questions/30091904/data-returned-by-google-api-is-not-up-to-date-real-time)

Reply:
That is a slightly different issue I believe in PR #58 . The bookmarked value (that which is saved in state and picked up from on the next sync) will be the last "golden" record however, we expect to always replicate data through the current date, even if there isn't data I'd expect us to replicate 0 or Null or whatever for a given metric. There is an important distinction between what replication-key values should be in a sync and what replication-key value should be used as a bookmark. Typically it is the maximum value that is used in the bookmark but here it is the maximum "golden" value, to account for continued aggregation on google's end. My goal for this ticket was to prove that it is not the tap that is filtering records, but the API that is not returning them. If you can provide proof of that, we can take this back to product and determine what is the proper result.

Reply:
@kspeer825 As per [Link1](https://stackoverflow.com/questions/48024517/ga-api-confirmation-to-get-full-day-of-data), [Link2](https://stackoverflow.com/questions/30091904/data-returned-by-google-api-is-not-up-to-date-real-time) data could take more than 24h to become golden data. Also, API does not return any rows if there are no records available for a particular date when the request is sent.
For example, if we request reports of last month on 22-12-2021(Current date). Then, as per the current tap code, the tap will make the API calls till the current date but will write only those records which are returned by API. So, if records of 22-12-2021 are not returned by API then the tap will write records till 21-12-2021(i.e. last record's timestamp). So, the replication-key value of the last record will be 21-12-2021 and which causes the failure of the existing assertion.

Here, below is a [snippet](https://github.com/singer-io/tap-google-analytics/blob/master/tap_google_analytics/sync.py#L139) of code that describes how bookmark will be written,

          if all_data_golden and not historically_syncing:
                    singer.write_bookmark(state,
                                          
                                          report["id"],
                                          report['profile_id'],
                                          {'last_report_date': report_date.strftime("%Y-%m-%d")})
                    singer.write_state(state)
                    if not is_data_golden and not historically_syncing:
                        # Stop bookmarking on first "isDataGolden": False
                        all_data_golden = False